src/tdsops.f90

```diff
@@ -0,0 +1,700 @@
+module m_tdsops
+   use m_common, only: dp, pi
+
+   implicit none
+
+   type :: tdsops_t
+      !! Tridiagonal Solver Operators class.
+      !!
+      !! Operator arrays are preprocessed in this class based on the arguments
+      !! provided. dist_fw and dist_bw are used in the first phase of the
+      !! distributed tridiagonal solver algorithm. dist_sa and dist_sc are used
+      !! in the final substutition phase. See the kernels_dist.f90 files in the
+      !! relevant backend folders.
+      !! coeff arrays define the specific rules of building the RHS
+      !! corresponding to the tridiagonal system to be solved, and used only in
+      !! the first phase of the distributed algorithm when building the RHS.
+      !! If a boundary condition is defined then coeffs_t and coeffs_b differ
+      !! from coeffs array and define the RHS rule for the first and last 4
+      !! entries in the tridiagonal system (n_halo = 4).
+      !!
+      !! This class does not know about the current rank or its relative
+      !! location among other ranks. All the operator arrays here are used when
+      !! executing a distributed tridiagonal solver phase one or two.
+      real(dp), allocatable, dimension(:) :: dist_fw, dist_bw, dist_sa, dist_sc
+      real(dp), allocatable :: coeffs(:), coeffs_t(:, :), coeffs_b(:, :)
+      integer :: n, n_halo
```

Comment:
Okay lets discuss n_halo in this thread

For the first, second, and staggared derivative and interpolation with classic scheme, n_halo=2 can be enough. However if the hyperviscosity is in use or optimized/aggressiveinterpolation schemes are selected, n_halo is 4. Because of this I implemented the universal derivative function assuming that n_halo is 4. This also leads RHS builder stencil to have 9 entries in total and it is also hardcoded in the universal function. If one initiates tdsops_t class with a n_halo other than 4, the existing distributed solver algorithm can't be used with the coeff arrays that tdsops_t provides when n_halo/=4.

However, the tdsops_t class is an independent class that can potentially be used with a kernel implementation that only supports n_halo=2 (or 3, or anything else if needed). Because it doesn't really have an impact on performance we don't need to bother having another implementation of distributed algorithm and use this with the schemes that require only two halo cells. But there is no harm implementing tdsops_t class in a way that supports an arbitrary n_halo number. Basically, thats why I included n_halo parameter here. We'll never instantiate this class with n_halo other than 4 (and therefore the default value for n_halo is 4 and its an optional parameter, so we don't even need to specify anything or consider this). And for some schemes, the halo number cannot be smaller than 2 or 4, so assignments such as coeff(4, :) = [...] is completely fine in my opinion. Maybe we can add conditional checks and warn people if they request incompatible n_halo value with some operations/schemes.

Reply:
I think it is misleading to have `n_halo` as a parameter / argument when under the hood, in this class and in some others, plenty of things are hard-coded for `n_halo=4`. At the very least, this should be clearly documented (in this class and in the universal derivative function) and the code should print some kind of warnings when n_halo is present and not 4.

Reply:
As a side remark, I am a bit surprised that the size of the stencil has no real impact on performance.